sam_view.c

```diff
@@ -68,13 +73,58 @@ typedef struct samview_settings {
     hts_filter_t *filter;
     int remove_flag;
     int add_flag;
+    auxhash_t remove_tag;
+    auxhash_t keep_tag;
 } samview_settings_t;
 
 
 // TODO Add declarations of these to a viable htslib or samtools header
 extern const char *bam_get_library(sam_hdr_t *header, const bam1_t *b);
 extern int bam_remove_B(bam1_t *b);
 
+// Copied from htslib/sam.c.
+// TODO: we need a proper interface to find the length of an aux tag,
+// or at the very make exportable versions of these in htslib.
+static inline int aux_type2size(uint8_t type)
+{
+    switch (type) {
+    case 'A': case 'c': case 'C':
+        return 1;
+    case 's': case 'S':
+        return 2;
+    case 'i': case 'I': case 'f':
+        return 4;
+    case 'd':
+        return 8;
+    case 'Z': case 'H': case 'B':
+        return type;
+    default:
+        return 0;
+    }
+}
+
+// Copied from htslib/sam.c.
+static inline uint8_t *skip_aux(uint8_t *s)
+{
+    int size = aux_type2size(*s); ++s; // skip type
+    uint32_t n;
+    switch (size) {
+    case 'Z':
+    case 'H':
+        while (*s) ++s;
+        return s + 1;
+    case 'B':
+        size = aux_type2size(*s); ++s;
+        memcpy(&n, s, 4); s += 4;
+        return s + size * n;
+    case 0:
+        abort();
```

Comment:
This is a bit of a harsh way of telling the user they have an invalid file.  Printing a suitable error message (preferably saying which read caused the problem) and calling exit would be nicer.

Reply:
This just comes from samtools/htslib@30b4033c and is just a demonstration of how ancient this PR is!

I'll copy the code from the latest htslib develop instead.